MSC3173: Expose stripped state events to any potential joiner (#3173)
matrix-org/matrix-doc: proposals/3173-expose-stripped-state-events.md

> is a mapping from room ID to room information. The room information is a mapping
> from a key `knock_state` to another mapping with key events being a list of
> `StrippedStateEvent`.

It is also provides an extended rationale of why this is useful:

> These stripped state events contain information about the room, most notably the
> room's name and avatar. A client will need this information to show a nice
> representation of pending knocked rooms. The recommended events to include are the
> join rules, canonical alias, avatar, name and encryption state of the room, rather
> than all room state. This behaviour matches the information sent to remote
> homeservers when remote users are invited to a room.

[MSC1772: Spaces](https://github.com/matrix-org/matrix-doc/pull/1772) additionally
recommends including the `m.room.create` event as one of the stripped state events:

> Join rules, invites and 3PID invites work as for a normal room, with the exception
> that `invite_state` sent along with invites should be amended to include the
> `m.room.create` event, to allow clients to discern whether an invite is to a
> space-room or not.

## Proposal

> Any user who is able to join a room can access the stripped state events of that room.

Reviewer: how?

Author: I think I've clarified this a bit to state that this MSC isn't adding a new mechanism for getting this information, but consolidating the definition and reasoning of why it is OK to expose this information.

Potential ways that a user might be able to join include, but are not limited to,
the following mechanisms:

* A room that has `join_rules` set to `public` or `knock`.
* A room that the user is in possession of an invite to (regardless of the `join_rules`).

Future MSCs might include additional mechanism for a user to join a room and
should consider this MSC, for example:

> * [MSC3083: Restricting room membership based on space membership](https://github.com/matrix-org/matrix-doc/pull/3083) proposes allowing users to join a room based on their membership in a space (as defined in [MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772)).

Reviewer: linebreaks at ~120ish characters please

Reviewer 2: Line breaks at 67-72 characters applied:
```suggestion
* [MSC3083: Restricting room membership based on space
 membership](https://github.com/matrix-org/matrix-doc/pull/3083) proposes 
 allowing users to join a room based on their membership  in a space (as 
 defined in [MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772)).
```
Signed-off by: Erkin Alp Güney <erkinalp9035@gmail.com>

Additionally, it is recommended, but not required, that homeserver implementations

> include the following as stripped state events:

Reviewer: to aid clarification, it might be worth trying to unify the two existing sets of stripped state as a "standardized set of events" or something so we don't have to write MSCs with 6 paragraphs (one for each place) every time we introduce a new kind of useful state event.

Author: That's pretty much what this is trying to do, while making it a "recommendation", not a "requirement". It is only recommended because privacy conscious people may want to trade off the user experience with not exposing information.

Do you have any suggestions of what would make this clearer?

Author: I think I've clarified this now!

* Create event (`m.room.create`)<sup id="a1">[2](#f1)</sup>
* Join rules (`m.room.join_rules`)
* Canonical alias (`m.room.canonical_alias`)
* Room avatar (`m.room.avatar`)
* Room name (`m.room.name`)

> * Encrypted status (`m.room.encryption`)<sup id="a2">[3](#f2)</sup>

Reviewer: Nitpick:
```suggestion
* Encryption information (`m.room.encryption`)<sup id="a2">[3](#f2)</sup>
```

"information" because the state event tells the client what algorithm to use, if any.

> * Encrypted status (`m.room.encryption`)<sup id="a2">[3](#f2)</sup>

Reviewer 3: It seems like the room topic (`m.room.topic`) is also a good one to include, especially because the Spaces summary list (at least in Element) includes the room topic.

On the other hand, it is more likely that sensitive information is stored in this field, so maybe not the best idea, and clients can just omit the description in these cases. It is currently [not recommended as one of the exposed state events in the `knock` endpoint](https://github.com/matrix-org/matrix-doc/pull/2403/files#diff-9dcacf9adb620c5346eba2f70f71bd50ed4f85a5e2844ac130e112165ce35c06R107-R109), but would be [exposed by the proposed spaces summary](https://github.com/matrix-org/matrix-doc/pull/2946/files#diff-892f7eda4445c4bf57fc6cad495c4328694cb3b6b7b3409092937223d50dfab0R60).

Author: Good catch, it is probably reasonable to expose this to someone who could join. This is also exposed for [public rooms](https://matrix.org/docs/spec/client_server/latest#get-matrix-client-r0-publicrooms) via the room directory. 👍

Author: I've included this now. 👍

Additionally, it is recommended, but not required, that homeserver implementations
include the following as stripped state events:

* Create event (`m.room.create`)<sup id="a2">[2](#f2)</sup>
* Join rules (`m.room.join_rules`)
* Canonical alias (`m.room.canonical_alias`)
* Room avatar (`m.room.avatar`)
* Room name (`m.room.name`)
* Encrypted status (`m.room.encryption`)<sup id="a3">[3](#f3)</sup>

This also implies that the above information is available to any potential joiner 
in the API proposed in [MSC2946: Spaces summary](https://github.com/matrix-org/matrix-doc/pull/2946). 
I.e. rooms which could be joined due to [MSC3083](https://github.com/matrix-org/matrix-doc/pull/3083)
can expose the information available in stripped state events.

## Potential issues

This is a generalization of current behavior and shouldn't introduce any new issues.

## Alternatives

A different approach to this would be to separately specify each situation in which
a user is allowed to see stripped state events, as we do currently for invites and
knocking.

> ## Security considerations

Reviewer 4: This allows accessing information without any trace. Previoously invites and knocks let the users in the room know that this information was shared. But now the user's server will share this information without the explicit action of someone in the room and without a trace being left.

I don't think this is an issue but I think it is worth pointing out all possible information leaks in this section.

Author: This would still be true for invite rooms, but you're right that this does open it up slightly for knocking. 👍 I'll include a note about that.

Author: I've moved some of this out of the footnote and into this section and expanded it! Thank you for the feedback!

The server-server API discussed in [MSC2946](https://github.com/matrix-org/matrix-doc/pull/2946)
does not know the user who is requesting a summary of the space, but should divulge
the above information if any member of a server could see it. It is up to the
calling server to properly filter this information.

Consider that Alice and Bob share a server; Alice is a member of a space, but Bob
is not. The remote server will not know whether the request is on behalf of Alice
or Bob (and hence whether it should share details of private rooms within that 
space).

Trust is placed in the calling server: if there are any users on the calling
server in the correct space, that calling server has a right to know about the
rooms in that space and should return the relevant summaries, along with enough 

> information that the calling server can then do some filtering.

Author: While implementing this I think it implies that the calling server needs to know whether access to a room requires access to a space (i.e. if the room is *not* world readable, but is joinable in some other way, such as is knockable or can be joined via MSC3083).

Author: @richvdh explained this a bit more succinctly than me:

* Bob is not a member of S1, but Alice is.
* neither is a member of R1
* Bob asks for state in R1
* the remote server returns the state because Alice is allowed to see it
* the local server doesn't know that Bob isn't allowed to see it

And came to the same conclusion:

> which I think means that the remote server needs to include the join restrictions in the response

Author: I've moved this to #3083, where it seems more appropriate.

> is a mapping from room ID to room information. The room information is a mapping
> from a key `knock_state` to another mapping with key events being a list of
> `StrippedStateEvent`.

It is also provides an extended rationale of why this is useful:

> These stripped state events contain information about the room, most notably the
> room's name and avatar. A client will need this information to show a nice
> representation of pending knocked rooms. The recommended events to include are the
> join rules, canonical alias, avatar, name and encryption state of the room, rather
> than all room state. This behaviour matches the information sent to remote
> homeservers when remote users are invited to a room.

[MSC1772: Spaces](https://github.com/matrix-org/matrix-doc/pull/1772) additionally
recommends including the `m.room.create` event as one of the stripped state events:

> Join rules, invites and 3PID invites work as for a normal room, with the exception
> that `invite_state` sent along with invites should be amended to include the
> `m.room.create` event, to allow clients to discern whether an invite is to a
> space-room or not.

## Proposal

> Any user who is able to join a room shall be allowed to have access the stripped
> state events of that room. No changes are proposed to the mechanics of how the
> users may get those state events, e.g. the `invite_state` of an invite or the
> roomd irectory.

Reviewer: ```suggestion
Any user who is able to join a room shall be allowed to have access the stripped
state events of that room. No changes are proposed to the mechanics of how the
users may get those state events, e.g. the `invite_state` of an invite or the
room directory.
```

> roomd irectory.

Reviewer: It's just really difficult to wrap my head around this MSC, sorry. It's not obviously clear what changes in the spec, if anything, which makes this questionable as an Matrix Spec Change proposal. If this proposal had an endpoint or something where the stripped state could be accessed, then there's material change to the spec. Alternatively, words similar to "unify the stripped state definitions" near the top here would also have material change, given they are divergent in the spec at the moment.

Author: Does the paragraph in the opening make it clearer?

> This MSC proposes allowing the stripped state events that are currently available
to invited and knocking users to any user who could potentially join a room. It
also consolidates the recommendation on which states events are available to
potential joiners.

Reviewer: ah, I missed that. It would be good to reinforce here in the opening statement of the proposal just so it's clear what to expect in the next few paragraphs (it will also have been a while between reading the intro and the proposal with the background section in the middle)

Author: I'll make them more consistent and re-iterate after the background. 👍

It is also proposed to create a single definition for what stripped state events
should be provided to be potential joiners. Thus, it is recommended (although not
required<sup id="a0">[0](#f0)</sup>) that homeserver implementations include the
following as stripped state events:

* Create event (`m.room.create`)<sup id="a1">[1](#f1)</sup>
* Join rules (`m.room.join_rules`)
* Canonical alias (`m.room.canonical_alias`)
* Room avatar (`m.room.avatar`)
* Room name (`m.room.name`)
* Encryption information (`m.room.encryption`)<sup id="a2">[2](#f2)</sup>
* Room topic (`m.room.topic`)<sup id="a3">[3](#f3)</sup>

## Potential issues

This is a generalization of current behavior and shouldn't introduce any new issues.

## Alternatives

A different approach to this would be to separately specify each situation in which
a user is allowed to see stripped state events, as we do currently for invites and
knocking.

## Security considerations

This would allow for invisibly accessing the stripped state of a room with `knock`
join rules. This is already trivially accessible by knocking on the room, but
currently users in the room would know that the knock occurred. This does not

> seem to be a major weakening of the security.

Author: This should probably cross reference the r[oom history visibility bits in the spec](https://spec.matrix.org/unstable/client-server-api/#room-history-visibility) which control how state is shared.

# MSC3173: Expose stripped state events to any potential joiner

It can be useful to view the partial state of a room before joining to allow a user
to know *what* they're joining. For example, it improves the user experience to
show the user the room name and avatar before joining.

It is already allowed to partially view the room state without being joined to
the room in some situations:

* If the room has `history_visibility: world_readable`, then anyone can inspect
  it (by calling `/state` on it).
* Rooms in the [room directory](https://matrix.org/docs/spec/client_server/latest#get-matrix-client-r0-publicrooms)
  expose some of their state publicly.
* [Invited users](https://matrix.org/docs/spec/server_server/r0.1.4#put-matrix-federation-v2-invite-roomid-eventid)
  and [knocking users](https://github.com/matrix-org/matrix-doc/pull/2403)
  receive stripped state events to display metadata to users.

This MSC proposes formalizing that the stripped state that is currently available

> to invited and knocking users is available to any user who could potentially join

Reviewer 5: Is “potentially join” recursive? I.e., if I'm in `A`, which allows joining `B`, which allows joining `C`, can I _potentially join_ `C`?

Reviewer: for the purposes of this MSC, no. Recursion like that would be determined in another access control MSC.

archived state. If the client joins the room then the current state will be given
as a delta against the archived state not the stripped state.

### Server behavior

It is recommended (although not required<sup id="a2">[2](#f2)</sup>) that
homeserver implementations include the following events as part of the stripped
state of a room:

* Create event (`m.room.create`)<sup id="a3">[3](#f3)</sup>
* Join rules (`m.room.join_rules`)
* Canonical alias (`m.room.canonical_alias`)
* Room avatar (`m.room.avatar`)
* Room name (`m.room.name`)
* Encryption information (`m.room.encryption`)<sup id="a4">[4](#f4)</sup>
* Room topic (`m.room.topic`)<sup id="a5">[5](#f5)</sup>

## Potential issues

This is a formalization of current behavior and should not introduce new issues.

## Alternatives

A different approach would be to continue with what is done today for invites,
knocking, the room directory: separately specify that a user is allowed to see
the stripped state (and what events the stripped state should contain).

## Security considerations

This would allow for invisibly accessing the stripped state of a room with `public`

> or `knock` join rules.

Reviewer 6: Is this still true, or can this section be removed now that the proposed API is removed?  I don't see that it currently introduces any new way for users to fetch the stripped state.

Author: I think this is somewhat still true since (in theory) a server could expose this information for a room which is knockable in e.g. the room directory (this is [somewhat covered by the knocking MSC already](https://github.com/Sorunome/matrix-doc/blob/soru/knock/proposals/2403-knock.md#changes-regarding-the-public-rooms-directory)).

be other pieces of information to include.

### Additional ways to access the stripped state

[MSC2946](https://github.com/matrix-org/matrix-doc/pull/2946) proposes including
the stripped state in the spaces summary. Not needing to rationalize what state
can be included for a potential joiner would simplify this (and future) MSCs.

### Additional ways to join a room

[MSC3083](https://github.com/matrix-org/matrix-doc/pull/3083) proposes a new
join rule due to membership in a space. This MSC would clarify that the stripped
state of a room is available to those joiners.

### Dedicated API for accessing the stripped state

Dedicated client-server and server-server APIs could be added to request the
stripped state events, but that is considered out-of-scope for the current
proposal.

This API would allow any potential joiner to query for the stripped state. If
the server does not know the room's state it would need to query other servers
for it.

## Unstable prefix

| Stable Endpoint | Unstable Endpoint |
|---|---|
| `/_matrix/client/r0/rooms/{roomIdOrAlias}/stripped_state` | `/_matrix/client/unstable/org.matrix.msc3173/rooms/{roomIdOrAlias}/stripped_state` |

> | `/_matrix/federation/v1/stripped_state/{roomId}` | `/_matrix/federation/unstable/org.matrix.msc3173/stripped_state/{roomId}` |

Reviewer: not needed now, I don't think

Author: Oh, yes. This should have been removed. Thank you!

It is already allowed to partially view the room state without being joined to
the room in some situations:

* If the room has `history_visibility: world_readable`, then anyone can inspect
  it (by calling `/state` on it).
* Rooms in the [room directory](https://matrix.org/docs/spec/client_server/latest#get-matrix-client-r0-publicrooms)
  expose some of their state publicly.
* [Invited users](https://matrix.org/docs/spec/server_server/r0.1.4#put-matrix-federation-v2-invite-roomid-eventid)
  and [knocking users](https://github.com/matrix-org/matrix-doc/pull/2403)
  receive stripped state events to display metadata to users.

This MSC proposes allowing the stripped state events that are currently available
to invited and knocking users to any user who could potentially join a room. It
also consolidates the recommendation on which events to include as stripped state
for potential joiners and provides a way to query for the stripped state directly.

This will allow for improved future use cases, such as:

* Improved user experience for more complicated access controls (e.g.
  [MSC3083](https://github.com/matrix-org/matrix-doc/pull/3083)).

> * More information available to platforms like matrix.to.

Author: @t3chguy I think you said you had a few other use-cases for it?

Reviewer 7: Join by alias in Element with including a preview of the target room
Showing a room preview in "Unable to preview" current peek failed screen

Author: Thanks! Hopefully I added this correctly...

* A room that the user is in possession of an invite to (regardless of the `join_rules`).

Future MSCs might include additional mechanism for a user to join a room and
should consider this MSC, for example:

* [MSC3083: Restricting room membership based on space membership](https://github.com/matrix-org/matrix-doc/pull/3083)
  proposes allowing users to join a room based on their membership in a space (as defined in
  [MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772)).

### Events to include in the stripped state

It is also proposed to create a single definition for what stripped state events
should be provided to be potential joiners. Thus, it is recommended (although not
required<sup id="a0">[0](#f0)</sup>) that homeserver implementations include the
following as stripped state events:

* Create event (`m.room.create`)<sup id="a1">[1](#f1)</sup>
* Join rules (`m.room.join_rules`)
* Canonical alias (`m.room.canonical_alias`)
* Room avatar (`m.room.avatar`)
* Room name (`m.room.name`)
* Encryption information (`m.room.encryption`)<sup id="a2">[2](#f2)</sup>
* Room topic (`m.room.topic`)<sup id="a3">[3](#f3)</sup>

> ### Stripped state API

Reviewer 8: is this thing actually implemented and tested? I think we'll need to see an impl before the MSC can be completed, since it's not quite as simple as it sounds (as evidenced by the thread below).

Reviewer 8: > I wonder if we can accelerate the MSC (for the benefit of the spaces work) by omitting the new API 

(I'm sorry if that makes you want to punch me, because someone else on the SCT insisted we needed an API)

Author: I think the original comment was about it being unclear what this MSC did without having an API. If this is now sufficiently clear we can leave that to a follow-up.

> is this thing actually implemented and tested?

No, but I hope to do that this week.

> `GET /_matrix/client/r0/rooms/{roomId}/stripped_state`

Reviewer 7: this API wouldn't work for rooms which the local server does not know as it does not allow specifying vias

Author: I think you're referring to `server_name` which is what [joining rooms can take](https://matrix.org/docs/spec/client_server/latest#post-matrix-client-r0-join-roomidoralias)? (Although spaces seems to use `via`...)

Good shout though, I guess we'll need to add a federation API for it too then.

Reviewer 7: Indeed, or the ability to specify a room alias but that isn't as flexible

Author: I guess this leads us to the decision about whether the requester should be included in the call over federation:

1. Similar to the logic used in [MSC3083](https://github.com/matrix-org/matrix-doc/blob/clokep/restricted-rooms/proposals/3083-restricted-rooms.md#discovery-of-restricted-rooms) we can leave this up to the calling server and provide a response as long as anyone on the calling server could see the stripped state.
2. Or the calling server can provide the requester and let the federated server decide whether the user has the proper permissions to see the state.

Author: > I guess this leads us to the decision about whether the requester should be included in the call over federation:

@turt2live and I discussed this a bit and it seems that this isn't a real concern since most of the time the data is only returned if it is publicly available. In other case (when it isn't publicly available) the requesting server should filter out, e.g. if the room is invite only then don't return info unless there's a pending invite for the user.

(There's some edge-cases here if the server does not return the join rules, but in that case the requesting server should default to `invite` and we "fail close".)

Author: We're going to defer the API endpoint to a follow-up.

be other pieces of information to include.

### Additional ways to access the stripped state

[MSC2946](https://github.com/matrix-org/matrix-doc/pull/2946) proposes including
the stripped state in the spaces summary. Not needing to rationalize what state
can be included for a potential joiner would simplify this (and future) MSCs.

### Additional ways to join a room

[MSC3083](https://github.com/matrix-org/matrix-doc/pull/3083) proposes a new
join rule due to membership in a space. This MSC would clarify that the stripped
state of a room is available to those joiners.

### Dedicated API for accessing the stripped state

Dedicated client-server and server-server APIs could be added to request the
stripped state events, but that is considered out-of-scope for the current
proposal.

This API would allow any potential joiner to query for the stripped state. If
the server does not know the room's state it would need to query other servers
for it.

## Unstable prefix

| Stable Endpoint | Unstable Endpoint |
|---|---|
| `/_matrix/client/r0/rooms/{roomIdOrAlias}/stripped_state` | `/_matrix/client/unstable/org.matrix.msc3173/rooms/{roomIdOrAlias}/stripped_state` |

> | `/_matrix/federation/v1/stripped_state/{roomId}` | `/_matrix/federation/unstable/org.matrix.msc3173/stripped_state/{roomId}` |

Reviewer 3: Does this still apply? It seems like this contradicts:

> Dedicated client-server and server-server APIs could be added to request the
> stripped state events, but that is considered out-of-scope for the current
> proposal.

from above.

Author: This was removed already.

Reviewer 3: Sorry! Must have been looking at a previous version or something. GitHub doesn't make that very obvious.
